File: Java/src/Kth_Largest_Element_in_an_Array.java
/**
 * Created by Kray on 2016/12/27.
 */
public class Kth_Largest_Element_in_an_Array {

    public static void main(String[] args) {
        Kth_Largest_Element_in_an_Array kth_largest_element_in_an_array = new Kth_Largest_Element_in_an_Array();
        System.out.println(kth_largest_element_in_an_array.findKthLargest(new int[]{1, 4, 2, 5, 3, 6}, 1));
    }

    public int findKthLargest(int[] nums, int k) {
        for (int i = nums.length - 1; i > 0; --i) {
            for (int j = 0; j < i; ++j) {
                if (nums[j + 1] < nums[j]) {
                    int temp = nums[j];
                    nums[j] = nums[j + 1];
                    nums[j + 1] = temp;
                }
            }
        }
        return nums[nums.length - k];
    }

}

File: Python/Kth Largest Element in an Array.py
from typing import List
import heapq


class Solution:

    def findKthLargest(self, nums: List[int], k: int) -> int:
        def partition(nums) -> int:
            pivot = nums[0]
            left, right = 1, 1
            while right < len(nums):
                if nums[right] >= pivot:
                    nums[left], nums[right] = nums[right], nums[left]
                    left += 1
                right += 1
            nums[0], nums[left - 1] = nums[left - 1], nums[0]
            return left - 1

        def quickselect(nums: List[int], k: int) -> int:
            if len(nums) == 1:
                return nums[0]
            idx = partition(nums)
            if idx < k - 1:
                return quickselect(nums[idx + 1:], k - idx - 1)
            elif idx > k - 1:
                return quickselect(nums[:idx], k)
            else:
                return nums[idx]

        return quickselect(nums, k)

    def findKthLargest_heap(self, nums: List[int], k: int) -> int:
        return heapq.nlargest(k, nums)[-1]
